Development documentation (#1079)
SUSE/kiwi: doc/source/development.rst

Create a local clone of the forked KIWI repository
--------------------------------------------------

.. code-block:: shell-session

    $ git clone https://github.com/YOUR-USERNAME/kiwi

    $ git remote add upstream https://github.com/SUSE/kiwi.git

Install required operating system packages
------------------------------------------

KIWI requires the following additional packages which are not provided by
:command:`pip`:

* XML processing libraries: libxml2 and libxslt (for :mod:`lxml`)
* Python header files (for :mod:`xattr`), usually provided by a package
  called `python-devel` or `python3-devel`

> * The `enchant` spell checking library

Author: I think this isn't a python package (it has python bindings though) and `:mod:` is for referring to python packages.

> * The `enchant` spell checking library

Reviewer: Oh you are right! Sorry. After reviewing all the text, my brain is melting now. :laughing:

* gcc compiler and glibc-devel header files
* The `ShellCheck <https://github.com/koalaman/shellcheck>`_ script
  linter.
* A manipulation program for ISO images, either: ``xorriso`` (preferred) or
  ``genisoimage``.

A full LaTeX installation is required to build the PDF documentation
[#f1]_.

On SUSE based systems run:

.. code-block:: shell-session

>     $ zypper install \

Author: Good idea! I'll try that if it installs everything with `--no-recommends`.

>     $ zypper install \

Author: That list was suspiciously short and missing **a lot** of packages.

>     $ zypper install \

Reviewer: Usually when I use this option, a lot of texlive packages are missing -- which was to be expected. :wink:  Not sure if it is useful to add the missing packages into the above line.

I thought it would be a nice idea, but if that doesn't work, just leave it as is. :-)

>     $ zypper install \

Author: I'd say let's keep it in since we already have a similar list for Fedora. It might be useful in case someone wants to build the pdf version of the documentation.

-----------------------------------------------

We recommend to setup a Python virtualenv for development. For Python 3
:command:`pyvenv` is used, as it is bundled with Python itself (see
https://docs.python.org/3.3/whatsnew/3.3.html#pep-405-virtual-environments
for details). For Python 2.7 use :command:`virtualenv`, which is provided
via pip or as an extra package by your favorite Linux distribution.

The following commands will setup an environment for Python 3:

.. code-block:: shell-session

   $ python3 -m venv .env3
   $ source .env3/bin/activate
   $ pip install -r .virtualenv.dev-requirements.txt
   $ python3 setup.py develop

The :command:`develop` target of the :command:`setup.py` script
automatically creates the application entry point called
:command:`kiwi-ng-3`, which allows you to call the your modified version of
KIWI from inside the virtual environment:

.. code-block:: shell-session

    $ kiwi-ng-3 --help

.. warning::

   The virtualenv's `$PATH` will not be taken into account when calling
   KIWI via :command:`sudo`! Use the absolute path to the KIWI executable
   to run an actual build using your local changes:

   .. code-block:: shell-session

>       $ /path/to/my/kiwi_local_copy/.env3/bin/kiwi-ng-3 system build ...

Reviewer: Isn't a `sudo` command missing here?

Or -- to simplify the line a bit -- you could use `$PWD` (assuming you are in the right directory):

```
$ sudo $PWD/.env3/bin/kiwi-ng-3 system build ...
```

> In order to leave the development mode kill your shell session or call:

Reviewer: Hmn... I'm not sure if we should advise the readers to "kill your shell session". The normal way is to deactivate the virtual environment, right? IMHO there should be only one, preferred way, not two.

.. code-block:: shell-session

    $ deactivate

To resume your work, :command:`cd` into your local Git repository and call:

.. code-block:: shell-session

    $ source .env3/bin/activate

Running the Unit Tests
----------------------

We use :command:`tox` to run the unit tests. Tox sets up its own
virtualenvs inside :file:`.tox` for multiple Python versions and should
thus **not** be invoked from inside your development virtualenv.

> Before submitting your changes via a Pull Request, ensure that all tests

Reviewer: ```suggestion
Before submitting your changes via a pull request, ensure that all tests
```

    $ tox

We also include `pytest-xdist` in the development virtualenv which allows
to run the unit tests in parallel. It is turned off by default but can be
enabled via:

.. code-block:: shell-session

    $ tox "-n NUMBER_OF_PROCESSES"

where you can insert an arbitrary number as `NUMBER_OF_PROCESSES` (or a
shell command like `$(nproc)`). Note that the double quotes around `-n
NUMBER_OF_PROCESSES` are required (otherwise :command:`tox` will consume
this command line flag instead of forwarding it to :command:`pytest`).

The previous call would run the unit tests for different Python versions,
check the source code for errors and build the documentation.

If you want to see the available targets, use the option `-l` to let
:command:`tox` print a list of them:

.. code-block:: shell-session

    $ tox -l
    check
    unit_py3_7
    unit_py3_6
    unit_py3_4
    unit_py2_7
    packagedoc

To only run a special target, use the `-e` option. The following
example runs the test cases for the Python 3.7 interpreter only:

.. code-block:: shell-session

    $ tox -e unit_py3_7

Create a Branch for each Feature or Bugfix
------------------------------------------

> Code changes should be done in an extra git branch. This allows for

Reviewer: ```suggestion
Code changes should be done in an extra Git branch. This allows for
```

    You can make multiple commits which is generally useful to
    give your changes a clear structure and to allow us to better
    review your work.

.. note::

>     Your work is important and should be signed to ensure the integrity of

Reviewer: "Should", same problem. See further explanation in line 338 (below).

    the repository and the code. Thus we recommend to setup a signing key
    as documented in :ref:`Signing_Git_Patches`.

.. code-block:: shell-session

    $ git commit -S -a

Run the tests and code style checks. All of these are also performed by the
`Travis CI <https://travis-ci.com/SUSE/kiwi>`_ and `GitLab CI
<https://gitlab.com/schaefi/kiwi-ci/pipelines>`_ integration test systems
when a pull request is created.

.. code-block:: shell-session

    $ tox

> Once everything is done push your local branch to the forked repository and

Reviewer: ```suggestion
Once everything is done, push your local branch to your forked repository and
```

create a pull request into the upstream repository.

.. code-block:: shell-session

    $ git push origin my-topic-branch

Thank you much for contributing to KIWI. Your time and work effort is very
much appreciated!

Coding Style
------------

KIWI follows the general PEP8 guidelines with the following exceptions:

- We do not use free functions at all. Even utility functions must be part
  of a class, but should be either prefixed with the `@classmethod` or
  `@staticmethod` decorators (whichever is more appropriate).

> - Constants are not named with all-caps.

Reviewer: Not sure about that:

```suggestion
- Constants are not named in all caps.
```

Author: I've used a different formulation.

- Do not set module and class level constants/variables, put these into the
  classes' `__init__` method.

> - We do not use type hints (yet) as the current code base maintains Python

Reviewer: ```suggestion
- We do not use type hints (yet) as the current code base needs to maintain Python
```

  2 compatibility.

Documentation
~~~~~~~~~~~~~

KIWI uses `Sphinx <https://www.sphinx-doc.org/en/master/>`_ for the API and
user documentation.

> In order to build the html documentation call:

Reviewer: ```suggestion
In order to build the HTML documentation call:
```

.. code-block:: shell-session

    tox -e doc

or to build the full documentation (including a PDF generated by LaTeX
[#f3]_):

.. code-block:: shell-session

    tox -e packagedoc

Document all your classes, methods, their parameters and their types using
the standard `reStructuredText
<https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html>`_
syntax as supported by Sphinx, an example class is documented as follows:

.. code-block:: python

   class Example(object):
       """
       **Example class**

       :param str param: A parameter
       :param bool : Source file name to compress
       :param list supported_zipper: List of supported compression tools
       :attr Optional[str] attr: A class attribute
       """
       def __init__(self, param, param_w_default=False):
           self.attr = param if param_w_default else None

       def method(self, param):
           """
           A method that takes a parameter.

           :param list param: a parameter
           :return: whether param is very long
           :rtype: bool
           """

>            return len(parame) > 50

Reviewer: ```suggestion
           return len(param) > 50
```

Try to stick to the following guidelines when documenting source code:

- Classes should be documented directly in their main docstring and not in
  `__init__`.

> - Document **every** parameter including their types.

Reviewer: Parameter may be a bit vague. You are probably referring to attributes, right? In this special case, public attributes?

```suggestion
- Document **every** public attribute including their types.
```

Author: I was actually thinking about function parameters, but public attributes should also be documented.

- Only public methods should be documented, private methods don't have to,
  unless they are complex and it is not easy to grasp what they do (which
  should be avoided anyway).

Please also document any user-facing changes that you implementing
(e.g. adding a new build type) in the user documentation, which can be
found in `doc/source`. General documentation should be put into the
`working_with_kiwi/` subfolder, whereas documentation about more
specialized topics would belong into the `building/` subfolder.

> Please adhere to a line limit of 75 characters when writing the user facing

Reviewer: One "please" is enough. :wink: 

```suggestion
Adhere to a line limit of 75 characters when writing the user facing
```

documentation [#f2]_.

Additional Information
----------------------

The following sections provides further information about the repository
integrity, version, package and documentation management.

.. _Signing_Git_Patches:

Signing Git Patches
~~~~~~~~~~~~~~~~~~~

To ensure the integrity of the repository and the code base, patches sent

> for inclusion should be signed with a GPG key.

Reviewer: Sorry to nitpick here, but the small word "should" kind of contradicts the whole integrity part.

I think, you need to be very clear what the project wants. Is it ok to accept patches which are not signed by a GPG key? Or are signed patches only a recommendation? Or do you *only* accept signed patches?

With *should* you indicate signing patches are optional. Was that your intention?

Author: cc @schaefi 

We **want** to have commits GPG signed, but have no mechanism in place that actually enforces this, thus I've changed the wording to describe the current state.

To prepare Git to sign commits, follow these instructions:

#. Create a key suitable for signing (it is not recommended to use
   existing keys to not mix it with your email environment):

   .. code-block:: shell-session

    $ gpg2 --expert --full-gen-key

#. Either choose a RSA key for signing (option `(4)`) or an ECC key for
   signing (option `(10)`). For a RSA key choose a key size of 4096 bits
   and for a ECC key choose Curve 25519 (option `(1)`). Enter a reasonable
   validity period (we recommend 2 to 5 years). Complete the key generation
   by entering your name and email address.

> #. Add the key ID to your git configuration

Reviewer: ```suggestion
#. Add the key ID to your Git configuration
```

   Add the following entries to :file:`~/.gitconfig` or :file:`.git/config`
   inside your repository (the latter will only apply to this repository)
   substituting the `11223344` with your signkey's ID:

   .. code:: ini

      [user]
              name = Joe Developer
              email = developer@foo.bar
              signingkey = 11223344

      [commit]
              gpgsign = true

   Or apply these settings via :command:`git config`:

   .. code-block:: shell-session

      $ git config --local user.signingkey $YOUR_SIGN_KEY_ID

>       $ git config --local commit.gpgSign true

Reviewer: Both ways are correct, but I would provide only **one** way to the user.
IMHO I would keep the code block with the git commands, but remove the editing part of the INI file.

   Omitting the flag `--local` will add these settings to
   :file:`~/.gitconfig` and make them global.

Bumping the Version
~~~~~~~~~~~~~~~~~~~

The KIWI project follows the `Semantic Versioning <https://semver.org>`_
scheme. We use the :command:`bumpversion` tool for consistent versioning.

Follow these instructions to bump the major, minor, or patch part of the

> KIWI version:

Reviewer: Before you run the `bumpversion` command, I would recommend to make sure you have a clean repo (= no modified or unknown files). If the repo is not clean, the command complains with an awful traceback.

*  For backwards-compatible bug fixes:

   .. code-block:: shell-session

    $ bumpversion patch

*  For additional functionality in a backwards-compatible manner. When
   changed, the patch level is reset to zero:

   .. code-block:: shell-session

    $ bumpversion minor

*  For incompatible API changes. When changed, the patch and minor
   levels are reset to zero:

   .. code-block:: shell-session

    $ bumpversion major

Creating a RPM Package
~~~~~~~~~~~~~~~~~~~~~~

We provide a template for a RPM spec file in
:file:`package/python-kiwi-spec-template` alongside with a rpmlint
configuration file and an automatically updated
:file:`python-kiwi.changes`.

To create the necessary files to build a RPM package via `rpmbuild`, run:

.. code-block:: shell-session

   $ make build

The sources are collected in the :file:`dist/` directory. These can be

> directly build it with :command:`rpmbuild`, :command:`fedpkg` or submitted

Reviewer: ```suggestion
directly build it with :command:`rpmbuild`, :command:`fedpkg`, or submitted
```

to the Open Build Service using :command:`osc`.

.. [#f1] Sphinx requires a plethora of additional LaTeX
         packages. Unfortunately there is currently no comprehensive list
         available. On Ubuntu/Debian installing `texlive-latex-extra`

>          should be sufficient, for Fedora please consult the package list

Reviewer: ```suggestion
         should be sufficient. For Fedora, consult the package list
```

         from :file:`.gitlab-ci.yml`.

.. [#f2] Configure your editor to automatically break lines and/or reformat
         paragraphs. For Emacs you can use `M-x set-fill-column RET 75` and
         `M-x auto-fill-mode RET`. For editing reStructuredText we

>          recommend `rst-mode` (built-in to Emacs since version `23.1`).

Reviewer: Do you have also recommendations for vi users? :wink:

Author: Not really, I've left the Cult of VIM a while ago and converted to the Church of Emacs. St. iGNUcius does not allow to have other deities beside him, so I must abstain from recommending anything to vi users. `</joking>`

Jokes aside: not really, I've never used vi properly. I'll try to dig something up, but a recommendation from an actual vi user is probably more helpful :wink:
